crates/sc-proof-of-time/src/pot_state.rs

```diff
@@ -0,0 +1,316 @@
+            Entry::Occupied(mut entry) => {
+                let proofs_for_slot = entry.get_mut();
+                // Reject if the sender already sent a proof for same slot number.
+                if proofs_for_slot.contains_key(&sender) {
+                    return Err(PotStateError::DuplicateProofFromPeer(sender));
+                }
+
+                // Reject if the there is an existing proof with same
+                // seed/key but different checkpoints.
+                if let Some(existing_proof) =
+                    proofs_for_slot.values().find(|p| p.seed == proof.seed)
+                {
+                    if existing_proof.checkpoints.as_slice() != proof.checkpoints.as_slice() {
+                        return Err(PotStateError::InvalidContextFromPeer(sender));
+                    }
+                }
+
+                // TODO: put a max limit on future proofs per slot number.
+                proofs_for_slot.insert(sender, proof.clone());
+                Ok(())
+            }
+        }
+    }
+
+    /// Called when the chain is extended with a new proof.
+    /// Tries to advance the tip as much as possible, by merging with
+    /// the pending future proofs.
+    fn merge_future_proofs(&self, state: &mut InternalState) {
+        loop {
+            let tip = if let Some(tip) = state.chain.last() {
+                tip.clone()
+            } else {
+                return;
+            };
+
+            // Get the pending proofs for (tip.slot_number + 1).
+            let next_slot = tip.slot_number + 1;
+            let proofs_for_slot = match state.future_proofs.get_mut(&next_slot) {
+                Some(proofs) => proofs,
+                None => return,
+            };
+
+            // Look for a proof with the matching key/seed.
+            let next_seed = tip.next_seed(None);
+            let next_key = tip.next_key();
+            let (sender, proof) = match proofs_for_slot
+                .iter()
+                .find(|(_, proof)| proof.seed == next_seed && proof.key == next_key)
+            {
+                Some((sender, proof)) => (*sender, proof.clone()),
+                None => return,
+            };
+
+            // Verify the proof.
+            if let Err(e) = self.proof_of_time.verify(&proof) {
+                warn!(target: LOG_TARGET, "pot state::merge future proofs: failed to verify: {e:?}");
+                proofs_for_slot.remove(&sender);
+                if proofs_for_slot.is_empty() {
+                    state.future_proofs.remove(&next_slot);
+                }
+                return;
+            }
+
+            // Extend the tip with the future proof.
+            self.add_to_tip(state, &proof, ExtendAction::NoMerge);
+        }
+    }
+
+    /// Adds the proof to the current tip
+    fn add_to_tip(&self, state: &mut InternalState, proof: &PotProof, action: ExtendAction) {
+        state.chain.push(proof.clone());
+        state.future_proofs.remove(&proof.slot_number);
```

Comment:
That pushes the clone to the callers, let us keep it localized

Reply:
It currently takes reference, your suggestion needs copy to be passed in

Reply:
That is the point! In idiomatic Rust if you need an owned value (and you do need one, otherwise you wouldn't clone it), then you need to take it as an argument.

Imagine you have a proof and call this method, but you don't care about the proof afterwards. You have two options: take a reference, clone it and drop original (1 clone) or give away the original and do no cloning at all.

This is really minor, just the way I think about things in Rust. If you have to clone something - you probably want an owned value in the first place.

Reply:
OK, done